### PR TITLE
[Permission] Rework - refactor 

### PR DESCRIPTION
**Pull request by lukmccall**
# Why

Part of #4972.

# How

- move logic responsible for asking/getting more than one permission to js side.
- extract base class for requesters
- remove unused functions
- use the macro to make `week and strong self`

# Test Plan

- test-suite passed.
- I tested it with a simple [demo](https://snack.expo.io/@lukaszkosmaty/permissions-refactoring) (run on several experiences to be able to test scoped permissions).

### CODE REVIEW

File: packages/expo-permissions/ios/EXPermissions.podspec [hunk 1]
```diff
@@ -3,22 +3,87 @@ require 'json'
...
 
-end
+  s.requires_arc    = true
+
+  s.default_subspec = "All"
+
+ s.subspec "Core" do |ss|
+    s.dependency 'UMCore'
+    s.dependency 'UMPermissionsInterface'
+    ss.source_files        = "EXPermissions/*.{h,m}"
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions.podspec]: I'm not a ruby expert, but from why these spaces?
```suggestion
    ss.source_files = "EXPermissions/*.{h,m}"
```

File: packages/expo-permissions/ios/EXPermissions/EXPermissionRequester.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+- (dispatch_queue_t)methodQueue;
+
+@end
+
+@interface EXPermissionBaseRequester : NSObject<EXPermissionRequester>
+
+@property (nonatomic, weak) id<EXPermissionRequesterDelegate> delegate;
+
+
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions/EXPermissionRequester.h]:
```suggestion
```

File: packages/expo-permissions/ios/EXPermissions/EXPermissionRequester.h [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+@end
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions/EXPermissionRequester.h]:
```suggestion
@end

```

File: packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m [hunk 4]
```diff
@@ -47,15 +41,9 @@ - (void)requestPermissionsWithResolver:(UMPromiseResolveBlock)resolve rejecter:(
 {
   [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m]:
```suggestion
  UM_WEAKIFY(self);
  [[AVAudioSession sharedInstance] requestRecordPermission:^(BOOL granted) {
```

File: packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m [hunk 5]
```diff
@@ -47,15 +41,9 @@ - (void)requestPermissionsWithResolver:(UMPromiseResolveBlock)resolve rejecter:(
...
     resolve([[self class] permissions]);
-    if (self->_delegate) {
-      [self->_delegate permissionRequesterDidFinish:self];
+    if (self.delegate) {
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m]:
```suggestion
	UM_ENSURE_STRONGIFY(self);
    if (self->_delegate) {
```

File: packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m [hunk 6]
```diff
@@ -47,15 +41,9 @@ - (void)requestPermissionsWithResolver:(UMPromiseResolveBlock)resolve rejecter:(
...
+      [self.delegate permissionRequesterDidFinish:self];
```
Comment (bbarthec) [inline packages/expo-permissions/ios/EXPermissions/Requesters/AudioRecording/EXAudioRecordingPermissionRequester.m]:
```suggestion
      [self->_delegate permissionRequesterDidFinish:self];
```

File: packages/expo-permissions/src/Permissions.ts [hunk 7]
```diff
@@ -30,25 +32,43 @@ export const REMINDERS = 'reminders';
...
+  return handlePermission(type);
 }
 
-async function _handlePermissionsRequestAsync(
+async function _handleMultiPermissionsRequestAsync(
   types: PermissionType[],
-  handlePermissions: (types: PermissionType[]) => Promise<PermissionMap>
-): Promise<PermissionResponse> {
+  handlePermission: (type: PermissionType) => Promise<PermissionInfo>
+): Promise<any> {
```
Comment (bbarthec) [inline packages/expo-permissions/src/Permissions.ts]: Definitely not.. 😭 
We do not want to smuggle any `any`s into our codebase 😉

Comment (bbarthec) [inline packages/expo-permissions/src/Permissions.ts]:
```suggestion
): Promise<PermissionResponse> {
```

File: packages/expo-permissions/src/Permissions.ts [hunk 8]
```diff
@@ -30,25 +32,43 @@ export const REMINDERS = 'reminders';
...
-    expires: coalesceExpirations(permissions),
-    permissions,
-  };
+  return await Promise.all(
+    types.map(async type => ({
+      [type]: await _handleSinglePermissionRequestAsync(type, handlePermission),
+    }))
+  )
+    .then(permissions =>
+      permissions.reduce((permission, acc) => {
```
Comment (bbarthec) [inline packages/expo-permissions/src/Permissions.ts]: Inlining is fun! 😄 
```suggestion
      permissions.reduce((permission, acc) => ({ ...acc, ...permission }), {})
```

File: packages/expo-permissions/src/Permissions.ts [hunk 9]
```diff
@@ -30,25 +32,43 @@ export const REMINDERS = 'reminders';
...
+        return { ...acc, ...permission };
+      }, {})
+    )
+    .then(permissions => {
```
Comment (bbarthec) [inline packages/expo-permissions/src/Permissions.ts]:
```suggestion
    .then(permissions => ({
        status: coalesceStatuses(permissions),
        expires: coalesceExpirations(permissions),
        permissions,
    })
```

**Review by bbarthec — COMMENTED** [review]
LGTM 💯 👍. Have changed any behaviour or just split the module into submodules and smoothed the flow?

**Comment by lukmccall** [comment]
@bbarthec, I only smoothed flow. From the users' perspective module should work like earlier.

**Pull request closed without merging**